src/perf/event_provider.cpp

```diff
@@ -0,0 +1,389 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2017,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <lo2s/config.hpp>
+#include <lo2s/log.hpp>
+#include <lo2s/perf/event_provider.hpp>
+
+#include <boost/filesystem.hpp>
+#include <regex>
+#include <sstream>
+
+extern "C" {
+#include <linux/perf_event.h>
+#include <linux/version.h>
+#include <syscall.h>
+#include <unistd.h>
+}
+
+namespace
+{
+#define PERF_EVENT(name, type, id)                                                                 \
+    {                                                                                              \
+        (name), (type), (id)                                                                       \
+    }
+#define PERF_EVENT_HW(name, id) PERF_EVENT(name, PERF_TYPE_HARDWARE, PERF_COUNT_HW_##id)
+#define PERF_EVENT_SW(name, id) PERF_EVENT(name, PERF_TYPE_SOFTWARE, PERF_COUNT_SW_##id)
+
+static const lo2s::platform::CounterDescription HW_EVENT_TABLE[] = {
+    PERF_EVENT_HW("cpu-cycles", CPU_CYCLES),
+    PERF_EVENT_HW("instructions", INSTRUCTIONS),
+    PERF_EVENT_HW("cache-references", CACHE_REFERENCES),
+    PERF_EVENT_HW("cache-misses", CACHE_MISSES),
+    PERF_EVENT_HW("branch-instructions", BRANCH_INSTRUCTIONS),
+    PERF_EVENT_HW("branch-misses", BRANCH_MISSES),
+    PERF_EVENT_HW("bus-cycles", BUS_CYCLES),
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 0, 0)
+    PERF_EVENT_HW("stalled-cycles-frontend", STALLED_CYCLES_FRONTEND),
+    PERF_EVENT_HW("stalled-cycles-backend", STALLED_CYCLES_BACKEND),
+#endif
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 3, 0)
```

Comment:
These checks are only there because they guard macros that aren't defined on lower kernel versions. If they weren't guarded, `lo2s` would fail to compile on those kernels.

Nonetheless, every event is checked at runtime to be available. If you compiled on a higher kernel version where these events were defined in a header and ran on a lower one, you'd get a `debug`-level warning (see [`event_is_openable`](https://github.com/tud-zih-energy/lo2s/blob/04912b89fd63c4f0a951c13fb0f465c933645006/src/perf/event_provider.cpp#L126)).